src/native.js

```diff
@@ -154,22 +154,36 @@ export function fireNativeTrackers(message, adObject) {
 export function getNativeTargeting(bid, bidReq) {
   let keyValues = {};
 
-  Object.keys(bid['native']).forEach(asset => {
-    const key = CONSTANTS.NATIVE_KEYS[asset];
-    let value = getAssetValue(bid['native'][asset]);
-
-    const sendPlaceholder = deepAccess(
-      bidReq,
-      `mediaTypes.native.${asset}.sendId`
-    );
-
-    if (sendPlaceholder) {
-      const placeholder = `${key}:${bid.adId}`;
-      value = placeholder;
-    }
+  if (deepAccess(bidReq, 'nativeParams.rendererUrl')) {
+    bid['native']['rendererUrl'] = getAssetValue(bidReq.nativeParams['rendererUrl']);
+  } else if (deepAccess(bidReq, 'nativeParams.adTemplate')) {
+    bid['native']['adTemplate'] = getAssetValue(bidReq.nativeParams['adTemplate']);
+  }
 
-    if (key && value) {
-      keyValues[key] = value;
+  Object.keys(bid['native']).forEach(asset => {
+    if (asset !== 'adTemplate') {
+      const key = CONSTANTS.NATIVE_KEYS[asset];
+      let value = getAssetValue(bid['native'][asset]);
+
+      const sendPlaceholder = deepAccess(
+        bidReq,
+        `mediaTypes.native.${asset}.sendId`
+      );
+
+      if (sendPlaceholder) {
+        const placeholder = `${key}:${bid.adId}`;
+        value = placeholder;
+      }
+
+      if (key && value) {
+        if (!deepAccess(bidReq, 'nativeParams.sendTargetingKeys') && typeof deepAccess(bidReq, 'nativeParams.sendTargetingKeys') !== 'undefined' && deepAccess(bidReq, 'nativeParams.' + asset + '.sendTargetingKeys')) {
+          keyValues[key] = value;
+        } else if (deepAccess(bidReq, 'nativeParams.sendTargetingKeys') && (typeof deepAccess(bidReq, 'nativeParams.' + asset + '.sendTargetingKeys') === 'undefined' || deepAccess(bidReq, 'nativeParams.' + asset + '.sendTargetingKeys'))) {
+          keyValues[key] = value;
+        } else if (typeof deepAccess(bidReq, 'nativeParams.sendTargetingKeys') === 'undefined' && (typeof deepAccess(bidReq, 'nativeParams.' + asset + '.sendTargetingKeys') === 'undefined' || deepAccess(bidReq, 'nativeParams.' + asset + '.sendTargetingKeys'))) {
+          keyValues[key] = value;
```

Comment:
I think this can be simplified a little.

```javascript
export function getNativeTargeting(bid, bidReq) {
  let keyValues = {};

  const globalSendTargetingKeys = deepAccess(
    bidReq,
    `mediaTypes.native.sendTargetingKeys`
  ) !== false; // if anything other than strict false, send keys

  Object.keys(bid['native']).forEach(asset => {
    const key = CONSTANTS.NATIVE_KEYS[asset];
    let value = getAssetValue(bid['native'][asset]);

    const sendPlaceholder = deepAccess(
      bidReq,
      `mediaTypes.native.${asset}.sendId`
    );

    const assetSendTargetingKeys = deepAccess(
      bidReq,
      `mediaTypes.native.${asset}.sendTargetingKeys`
    );

    const sendTargeting = typeof assetSendTargetingKeys === 'boolean' ? assetSendTargetingKeys : globalSendTargetingKeys;

    if (sendPlaceholder) {
      const placeholder = `${key}:${bid.adId}`;
      value = placeholder;
    }

    if (key && value && sendTargeting) {
      keyValues[key] = value;
    }
  });

  return keyValues;
}
```

before the loop we determine what the "top level" setting is -> If anything other than boolean false, then it is true, otherwise false.

Then for each asset, if sendTargetingKeys is a boolean for that asset, use that value, else use the top level one.

So instead of the if else if's, just add the boolean check in the if(key and value).

Reply:
I did not test this, but It should work in theory, I may have a typo or something.

Let me know if you have questions or think it does not work!

Reply:
second @robertrmartinez 's solution. And not only this version cleaner, also that version has logic issue. I believe we suppose keep existing structure `mediaType.native`, but the code is looking for `nativeParams `, which is a new structure.

Reply:
If I am not wrong, the solution misses a use case for `mediaType.native.keysToSend`.

Reply:
In the consolidated spec we left this out, as the idea was the most often use cases will be send all or none.

I believe others agreed, but if everyone else thinks so, we can add it in.

I would rather there be only one or the other though.

Either each asset can have the `sendTargetingKeys` or at top level there is a `keysToSend` 

I do not think it makes sense to have the code written to handle both.

Reply:
This logic is much cleaner thanks. Pushed an update for this.